### PR TITLE
Add support for stateless code flow

### PR DESCRIPTION
**Pull request by smalihaider**
Added the capability of stateless code flow. If stateless flag is set to true, nothing is stored in the storage. The code and access token is embedded with the user_info which is required later for the evaluation of required claims.

### CODE REVIEW

File: src/pyop/authz_state.py [hunk 1]
```diff
@@ -78,9 +82,11 @@ def __init__(self, subject_identifier_factory, authorization_code_db=None, acces
         Mapping of user id's to subject identifiers.
         """
         self.subject_identifiers = subject_identifier_db if subject_identifier_db is not None else {}
+        self.stateless = stateless
 
-    def create_authorization_code(self, authorization_request, subject_identifier, scope=None):
-        # type: (AuthorizationRequest, str, Optional[List[str]]) -> str
+    def create_authorization_code(self, authorization_request, subject_identifier, scope=None, user_id=None,
+                                  user_info=None):
```
Comment (c00kiemon5ter) [inline src/pyop/authz_state.py:88]: What does `user_id` and `user_info` hold?

`user_id` and `subject_identifier` look the same.

File: src/pyop/provider.py [hunk 2]
```diff
@@ -57,9 +59,11 @@ def __init__(self, signing_key, configuration_information, authz_state, clients,
         :param clients: see <a href="https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata">
             "OpenID Connect Dynamic Client Registration 1.0", Section 2</a>
         :param userinfo: read-only interface for user info
+        :param stateless: users, authorization codes, and tokens related data is not persisted
         :param id_token_lifetime: how long the signed ID Tokens should be valid (in seconds), defaults to 1 hour
         """
         self.signing_key = signing_key
+        self.encryption_key = encryption_key
```
Comment (c00kiemon5ter) [inline src/pyop/provider.py:66]: `encryption_key` doesn't seem to be used by the `Provider`

Comment (smalihaider) [reply src/pyop/provider.py]: removed

File: src/pyop/authz_state.py (hunk 1 above)
Comment (smalihaider) [reply src/pyop/authz_state.py:88]: - user_id is a local user identifier which means that it is also the key used to look up the user_db.
- user_info is that user_db holding all the user_claims
- I debugged and indeed the value of user_id and subject_identifier was the same.

I have made the respective changes accordingly.

File: src/pyop/provider.py [hunk 3]
```diff
@@ -180,12 +197,25 @@ def authorize(self, authentication_request,  # type: AuthorizationRequest
...
+            logger.debug(
+                'issued id_token=%s from requested_claims=%s userinfo=%s extra_claims=%s',
+                response['id_token'],
+                requested_claims,
+                user_claims,
+                extra_id_token_claims,
+            )
+
+        if self.stateless:
+            del self.userinfo[user_id]
```
Comment (c00kiemon5ter) [inline src/pyop/provider.py:218]: @smalihaider Why do we need to remove the entry at this point?

There is an assumption that pyop does not modify the user-db. Can we do this on the satosa-frontend level? It is on the satosa-frontend level we also add entries to that db.

File: src/pyop/userinfo.py [hunk 4]
```diff
@@ -18,16 +18,22 @@ def __getitem__(self, item):
     def __contains__(self, item):
         return item in self._db
 
-    def get_claims_for(self, user_id, requested_claims):
+    def __delitem__(self, key):
+        del self._db[key]
```
Comment (c00kiemon5ter) [inline src/pyop/userinfo.py:22]: _note_: if https://github.com/IdentityPython/pyop/pull/41/files?diff=unified&w=1#r852268477 is removed, this should also be removed.

File: src/pyop/provider.py (hunk 3 above)
Comment (smalihaider) [reply src/pyop/provider.py]: I agree, change have been pushed.

File: src/pyop/userinfo.py (hunk 4 above)
Comment (smalihaider) [reply src/pyop/userinfo.py]: agreed, removed.

**Pull request merged**